pkg/kubelet/cm/cpumanager/cpu_assignment.go

```diff
@@ -518,6 +577,45 @@ func (a *cpuAccumulator) takeFullSockets() {
 		a.take(cpusInSocket)
 	}
 }
+func (a *cpuAccumulator) takeFullUnCore() {
+	for _, uncore := range a.freeUnCoreCache() {
+		cpusInUnCore := a.topo.CPUDetails.CPUsInUnCoreCaches(uncore)
+		if !a.needsAtLeast(cpusInUnCore.Size()) {
+			continue
+		}
+		klog.V(4).InfoS("takeFullCCD: claiming CCD", "uncore", uncore)
+		a.take(cpusInUnCore)
+	}
+}
+
+// First try to take partial uncorecache (CCD), if available and the request size can fit w/in the uncorecache.
+// Second try to take the full CCD if available and need is at least the size of the uncorecache group.
+func (a *cpuAccumulator) takeUnCoreCache() {
+	// check if SMT ON
+
+	for _, uncore := range a.allUnCoreCache() {
+		numCoresNeeded := a.numCPUsNeeded / a.topo.CPUsPerCore() // this is another new change
+
+		var freeCPUsInUncorecache cpuset.CPUSet
+		// need to get needed cores in uncorecache
+		freeCoresInUncorecache := a.details.CoresNeededInUnCoreCache(numCoresNeeded, uncore)
+		klog.V(2).InfoS("free cores from a.details list: ", "freeCoresInUncorecache", freeCoresInUncorecache)
+		for _, coreID := range freeCoresInUncorecache.List() {
+			freeCPUsInUncorecache = freeCPUsInUncorecache.Union(a.topo.CPUDetails.CPUsInCores(coreID))
+		}
+		klog.V(2).InfoS("freeCPUsInUncorecache  : ", "freeCPUsInUncorecache", freeCPUsInUncorecache)
+		if a.numCPUsNeeded == freeCPUsInUncorecache.Size() {
+			klog.V(4).InfoS("takePartialUncore: claiming cores from Uncorecache ID", "uncore", uncore)
+			a.take(freeCPUsInUncorecache)
+		}
+		// take full Uncorecache if the numCPUsNeeded is greater the L3 cache size
+		a.takeFullUnCore()
+
+		if a.isSatisfied() {
+			return
+		}
+	}
```

Comment:
is this logic matching the intent described in the KEP? IIRC we should take full uncore group first, then try to allocate the reminder and the code as I read it seems to be doing the opposite

Reply:
It seems it doesn't: https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/4800-cpumanager-split-uncorecache#design-details
```

    If the required CPUs are greater than or equal the size of the total amount of CPUs to an uncore cache, assign the full uncore cache CPUs if they are all unscheduled within the group. Subtract the amount of scheduled CPUs from the quantity of required CPUs.
        Continue scanning uncore caches in numerical order and assigning full uncore cache CPU groups until the required quantity is less than the total number of CPUs to a uncore cache group.
    If the required CPUs are less than the size of the total amount of CPUs to an uncore cache, scan through each uncore cache index in numerical order starting from the first index, and assign CPUs if there are enough available CPUs within the uncore cache group.
```

I'd expect to call first `takeFullUnCore()` then try to do partial allocation if cores are still needed

Reply:
Overall, I finally realized that the code does what the KEP describes, which is good ofc. What we can and should do is to gather feedback from early user and review our algorithms first in the KEP and then in the code, tuning as needed. We can do this staying in alpha in 1.33 and postponing the beta graduation.